core/src/main/java/org/apache/struts2/components/Debug.java

```diff
@@ -50,31 +50,46 @@ public Debug(ValueStack stack, HttpServletRequest request, HttpServletResponse r
     public void setReflectionProvider(ReflectionProvider prov) {
         this.reflectionProvider = prov;
     }
-    
+
     protected String getDefaultTemplate() {
         return TEMPLATE;
     }
 
     public boolean start(Writer writer) {
         boolean result = super.start(writer);
 
-        ValueStack stack = getStack();
-        Iterator iter = stack.getRoot().iterator();
-        List stackValues = new ArrayList(stack.getRoot().size());
-        while (iter.hasNext()) {
-            Object o = iter.next();
-            Map values;
-            try {
-                values = reflectionProvider.getBeanMap(o);
-            } catch (Exception e) {
-                throw new StrutsException("Caught an exception while getting the property values of " + o, e);
+        if (showDebug()) {
+            ValueStack stack = getStack();
+            Iterator iter = stack.getRoot().iterator();
+            List stackValues = new ArrayList(stack.getRoot().size());
+            while (iter.hasNext()) {
+                Object o = iter.next();
+                Map values;
+                try {
+                    values = reflectionProvider.getBeanMap(o);
+                } catch (Exception e) {
+                    throw new StrutsException("Caught an exception while getting the property values of " + o, e);
+                }
+                stackValues.add(new DebugMapEntry(o.getClass().getName(), values));
             }
-            stackValues.add(new DebugMapEntry(o.getClass().getName(), values));
+
+            addParameter("stackValues", stackValues);
         }
+        return result;
+    }
 
-        addParameter("stackValues", stackValues);
+    @Override
+    public boolean end(Writer writer, String body) {
+        if (showDebug()) {
+            return super.end(writer, body);
+        } else {
+            popComponentStack();
+            return false;
+        }
+    }
 
-        return result;
+    protected boolean showDebug() {
+        return (devMode || Boolean.TRUE == PrepareOperations.getDevModeOverride());
```

Comment:
Ah! So you wanted to have an option to enable the `<s:debug/>` on production that's why you had used the `disabled` property. In such case `getDevModeOverride` is a better choice 👍

Reply:
yes, or more specifically for the Struts Showcase. 99 percent of users will not use it

Reply:
👍